java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java

```diff
@@ -0,0 +1,101 @@
+package org.apache.arrow.memory;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Reference Manager manages one or more ArrowBufs that share the
+ * reference count for the underlying memory chunk.
+ */
+public interface ReferenceManager {
+
+  /**
+   * Return the reference count
+   * @return reference count
+   */
+  int getRefCount();
+
+  /**
+   * Decrement this reference manager's reference count for the associated underlying
+   * memory. If the reference count drops to 0, it implies that ArrowBufs managed by this
+   * reference manager no longer need access to the underlying memory
+   * @return the new reference count
+   */
+  boolean release();
+
+  /**
+   * Decrement this reference manager's reference count for the associated underlying
+   * memory. If the reference count drops to 0, it implies that ArrowBufs managed by this
+   * reference manager no longer need access to the underlying memory
+   * @param increment the count to increase the reference count by
+   * @return the new reference count
+   */
+  boolean release(int increment);
+
+  /**
+   * Increment this reference manager's reference count by 1 for the associated underlying
+   * memory.
+   */
+  void retain();
+
+  /**
+   * Increment this reference manager's reference count by a given amount for the
+   * associated underlying memory
+   * @param increment the count to increase the reference count by
+   */
+  void retain(int increment);
+
+  /**
+   * Create a new ArrowBuf that is associated with an alternative allocator for the purposes of
+   * memory ownership and accounting. This has no impact on the reference counting for the current
+   * ArrowBuf except in the situation where the passed in Allocator is the same as the current buffer.
+   *
+   * This operation has no impact on the reference count of this ArrowBuf. The newly created
+   * ArrowBuf with either have a reference count of 1 (in the case that this is the first time this
+   * memory is being associated with the target allocator or in other words allocation manager currently
+   * doesn't hold a mapping for the target allocator) or the current value of the reference count for
+   * the target allocator-reference manager combination + 1 in the case that the provided allocator
+   * already had an association to this underlying memory.
+   *
+   * @param srcBuffer source ArrowBuf
+   * @param targetAllocator The target allocator to create an association with.
+   * @return A new ArrowBuf which shares the same underlying memory as this ArrowBuf.
+   */
+  ArrowBuf retain (ArrowBuf srcBuffer, BufferAllocator targetAllocator);
```

Comment:
this name doesn't seem to match up well with the documentation if an actual arrowbuf is being allocated.